### PR TITLE
CLN: dataframe.io

### PR DESCRIPTION
**Pull request by sinhrks**
Cleaned up ``dataframe.io`` related files and updated API doc.

### CODE REVIEW

**Comment by mrocklin** [comment]
In general this refactoring seems like a good idea to me

**Comment by mrocklin** [comment]
cc @jcrist

File: docs/source/dataframe-api.rst [hunk 1]
```diff
@@ -164,9 +164,13 @@ Other functions
 .. currentmodule:: dask.dataframe
 
 .. autofunction:: read_csv
+.. autofunction:: read_hdf
 .. autofunction:: from_array
 .. autofunction:: from_pandas
 .. autofunction:: from_bcolz
+.. autofunction:: from_dask_array
+.. autofunction:: from_castra
```
Comment (sinhrks) [inline docs/source/dataframe-api.rst]: ok. I also removed existing ``to_castra`` from API doc.

**Comment by mrocklin** [comment]
The Py27 test failure on `dask/bag/tests/test_bag.py::test_groupby_with_indexer` is somewhat ominous here.  Possibly related to recent `partd` fixes?

**Comment by mrocklin** [comment]
+1 from me

**Comment by mrocklin** [comment]
This has merge conflicts from recently merged PRs.

**Comment by jcrist** [comment]
`dask.dataframe.io` needs to be added to packages in setup.py: https://github.com/dask/dask/blob/master/setup.py#L17. Otherwise we'll get a repeat of this issue: https://github.com/dask/dask/issues/1067.

Otherwise, this looks good to me.

**Pull request merged**